platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/heartbeats/HeartbeatProtocol.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.heartbeats;
+
+import com.swirlds.base.ArgumentUtils;
+import com.swirlds.common.system.NodeId;
+import com.swirlds.platform.Connection;
+import com.swirlds.platform.network.ByteConstants;
+import com.swirlds.platform.network.NetworkMetrics;
+import com.swirlds.platform.network.NetworkProtocolException;
+import com.swirlds.platform.network.protocol.Protocol;
+import com.swirlds.platform.network.unidirectional.UnidirectionalProtocols;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+
+/**
+ * Sends a heartbeat to the other node and measures the time it takes to receive a response.
+ */
+public class HeartbeatProtocol implements Protocol {
```

Comment:
Once you switch over to using a `Time` object to simulate time I think a unit test for this class would be nice to have. Take a look at the implementations of `Connection`, there should be some that would let you simulate two heartbeat protocols talking to each other.

When you write these tests, I'd try to test this class directly without spinning up a negotiator object (for the sake of simplicity).

Reply:
unit tests written